homeassistant/auth/permissions/merge.py

```diff
@@ -35,33 +40,29 @@ def _merge_policies(sources: list[CategoryType]) -> CategoryType:
     # If there are multiple sources with a dict as policy, we recursively
     # merge each key in the source.
 
-    policy: CategoryType = None
-    seen: set[str] = set()
-    for source in sources:
-        if source is None:
-            continue
+    def merge_dicts(dicts: List[Dict[str, CategoryType]]) -> Dict[str, CategoryType]:
+        merged_dict: Dict[str, CategoryType] = {}
+        seen: set[str] = set()
+        for dictionary in dicts:
+            for key, value in dictionary.items():
+                if key not in seen:
+                    seen.add(key)
+                    if key in merged_dict and isinstance(value, dict):
+                        # Recursively merge sub-dictionaries
+                        merged_dict[key] = merge_dicts(
+                            [d.get(key, {}) for d in dicts if isinstance(d, dict)]
```

Comment:
Test `test_merging_permissions_true_rules_dict()` is broken because of this line, I think: if `d` is a `bool`, its value is discarded while it should override the value. Looking at the previous implementation, it seems that a value of `True` should always win. I don't think your reworked version keeps this behavior.

Note that reversing policy1 and policy2 in the call to `merge_policies` make the test pass, which is a weakness of the test. I would recommend to improve the test to make sure the operation is commutative.